src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.util;
+
+import static org.wildfly.security.util.ProviderUtil.INSTALLED_PROVIDERS;
+
+import org.wildfly.security.WildFlyElytronProvider;
+import org.wildfly.security.util._private.Arrays2;
+
+import java.security.Provider;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+
+/**
+ * A supplier which uses a service loader to find all {@link Provider} instances that aren't in the list of
+ * installed security providers and returns them as an array. The result is then cached.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ * @since 1.2.2
+ */
+public class ProviderServiceLoaderSupplier extends ServiceLoaderSupplier<Provider> {
+
+    private static final HashSet INSTALLED_PROVIDERS_SET;
+    static {
+        Provider[] providers = INSTALLED_PROVIDERS.get();
+        INSTALLED_PROVIDERS_SET = new HashSet((providers != null ? providers.length : 0) + 1);
+        INSTALLED_PROVIDERS_SET.add(WildFlyElytronProvider.class);
+        if (providers != null) {
+            for (int i = 0; i < providers.length; i++) {
+                INSTALLED_PROVIDERS_SET.add(providers[i].getClass());
+            }
+        }
+    }
+
+    public ProviderServiceLoaderSupplier(final ClassLoader classLoader) {
+        super(Provider.class, classLoader);
+    }
+
+    protected Provider[] loadServices(final Class<Provider> service, final ClassLoader classLoader) {
+        ArrayList<Provider> list = new ArrayList<>();
+        ServiceLoader<Provider> loader = ServiceLoader.load(service, classLoader);
+        Iterator<Provider> iterator = loader.iterator();
+        for (;;) try {
+            if (! iterator.hasNext()) {
+                return list.toArray(Arrays2.createArray(service, list.size()));
```

Comment:
I don't remember if @darranl has a different recommendation here, but AFAIK accepted best practice is to pass in a constant 0-sized array to ``Collection.<T>toArray(T[] array)`` methods.

Reply:
This is actually just a copy/paste from ```ServiceLoaderSupplier#loadServices```. This isn't always 0-sized, this is the result after iterating over the providers.

Reply:
``ServiceLoaderSupplier`` is generic; that's why it uses this array creation strategy.  But since this class is not generic, it does not need to use the dynamic creation strategy.  That said, this isn't incorrect, it's just non-optimal.

Reply:
Ah, yes, that was silly :) I'll fix that.